python/ray/dashboard/modules/node/node_head.py

```diff
@@ -170,104 +157,122 @@ def get_internal_states(self):
-                            f"{dashboard_consts.DASHBOARD_AGENT_PORT_PREFIX}"
-                            f"{node_id}"
+                published = await subscriber.poll(batch_size=200)
+                for node_id, node_info in published:
+                    if node_id is not None:
+                        yield await get_or_create_event_loop().run_in_executor(
+                            self._dashboard_head._thread_pool_executor,
+                            gcs_node_info_to_dict,
+                            node_info,
                         )
-                        agent_port = await self._gcs_aio_client.internal_kv_get(
-                            key.encode(),
-                            namespace=ray_constants.KV_NAMESPACE_DASHBOARD,
-                            timeout=GCS_RPC_TIMEOUT_SECONDS,
-                        )
-                        if agent_port:
-                            agents[node_id] = json.loads(agent_port)
-                for node_id in agents.keys() - set(alive_node_ids):
-                    agents.pop(node_id, None)
-
-                DataSource.agents.reset(agents)
-                DataSource.nodes.reset(nodes)
+                    # yield control to APIs
+                    await asyncio.sleep(0)
             except Exception:
-                logger.exception("Error updating nodes.")
-            finally:
-                self._node_update_cnt += 1
-                # _head_node_registration_time_s == None if head node is not
-                # registered.
-                head_node_not_registered = not self._head_node_registration_time_s
-                # Until the head node is registered, we update the
-                # node status more frequently.
-                # If the head node is not updated after 10 seconds, it just stops
-                # doing frequent update to avoid unexpected edge case.
+                logger.exception("Error updating nodes from subscriber.")
+
+    async def _update_node(self, node: dict):
+        node_id = node["nodeId"]  # hex
+        if node["isHeadNode"] and not self._head_node_registration_time_s:
+            self._head_node_registration_time_s = time.time() - self._module_start_time
+            # Put head node ID in the internal KV to be read by JobAgent.
+            # TODO(architkulkarni): Remove once State API exposes which
+            # node is the head node.
+            await self._gcs_aio_client.internal_kv_put(
+                ray_constants.KV_HEAD_NODE_ID_KEY,
+                node_id.encode(),
+                overwrite=True,
+                namespace=ray_constants.KV_NAMESPACE_JOB,
+                timeout=GCS_RPC_TIMEOUT_SECONDS,
+            )
+        assert node["state"] in ["ALIVE", "DEAD"]
+        is_alive = node["state"] == "ALIVE"
+        # prepare agents for alive node, and pop for dead node
+        if is_alive:
+            if node_id not in DataSource.agents:
+                asyncio.create_task(self._update_agent(node_id))
```

Comment:
Let's not optimize what doesn't need to be optimized -- just await in here

Reply:
Ok, i see what you're trying to do here -- add a comment explaining why it needs to be async

Reply:
added comment.